scripts/src/qc_table_functions.R

```diff
@@ -0,0 +1,374 @@
+#'
+#' @param cell_set_meta A data frame containing metadata for cell sets.
+#' @param cell_line_cols A character vector specifying the column names that define a unique cell line.
+#'
+#' @return A data frame with one row per cell_set and a column `n_expected_lines` indicating the number of unique expected cell lines in each set.
+#'
+#' @import dplyr
+compute_expected_lines <- function(cell_set_meta, cell_line_cols) {
+  # Get number of expected cell lines for each cell_set
+  cell_set_meta %>%
+    dplyr::group_by(across(all_of(cell_line_cols))) %>%
+    dplyr::filter(n() == 1) %>%
+    dplyr::ungroup() %>%
+    dplyr::group_by(cell_set) %>%
+    dplyr::summarise(
+      n_expected_lines = dplyr::n_distinct(across(all_of(cell_line_cols))), # Count unique cell_lines for each cell_set
+    ) %>%
+    dplyr::ungroup()
+}
+
+#' Compute read stats
+#'
+#' This function calculates summary statistics related to reads and cell line recovery for annotated counts data,
+#' combining metadata and performing group-wise computations.
+#'
+#' @param annotated_counts A data frame containing read data with annotations for barcodes, cell lines, and counts.
+#' @param cell_set_meta A data frame containing metadata about cell sets and their expected cell lines.
+#' @param group_cols A character vector specifying the grouping columns (default: `c("pcr_plate", "pcr_well")`).
+#' @param metric A string indicating the column name of the metric to use for calculations (default: `"n"`).
+#' @param cell_line_cols A character vector specifying the column names that define a unique cell line (default: `c("depmap_id", "pool_id")`).
+#' @param count_threshold An integer specifying the minimum count threshold for a cell line to be considered recovered (default: `40`).
+#'
+#' @return A data frame summarizing read statistics for each group, including total reads, expected reads, control barcode reads,
+#' recovered cell lines, and their fractions.
+#'
+#' @import dplyr
+compute_read_stats <- function(annotated_counts, cell_set_meta, group_cols = c("pcr_plate","pcr_well"),
+                               metric = "n", cell_line_cols = c("depmap_id", "pool_id"), count_threshold = 40) {
+  # Compute expected lines from cell_set_meta
+  expected_lines <- compute_expected_lines(cell_set_meta, cell_line_cols)
+
+  result <- annotated_counts %>%
+    dplyr::left_join(expected_lines, by = "cell_set") %>% # Add n_expected_lines from lookup
+    dplyr::group_by(across(all_of(group_cols))) %>%
+    dplyr::summarise(
+      # Total reads
+      n_total_reads = sum(.data[[metric]], na.rm = TRUE),
+      # Reads mapping to cell lines
+      n_bc_reads = sum(.data[[metric]][expected_read], na.rm = TRUE),
+      # Fraction of reads mapping to cell lines
+      fraction_expected_reads = n_bc_reads / n_total_reads,
+      # Reads mapping to control barcodes
+      n_cb_reads = sum(!is.na(cb_name) & cb_name != "", na.rm = TRUE),
```

Comment:
Is this trying to compute the number of CBCs recovered? If so, would be good to have a count threshold here. And also would be good to change the name to n_cb_recovered.

If you want to compute the number of reads mapping to CBCs instead, then would need to sum the counts in this statement